vault/acl.go

```diff
@@ -47,25 +47,83 @@ func NewACL(policies []*Policy) (*ACL, error) {
 			// Check for an existing policy
 			raw, ok := tree.Get(pc.Prefix)
 			if !ok {
-				tree.Insert(pc.Prefix, pc.CapabilitiesBitmap)
+				tree.Insert(pc.Prefix, pc.Permissions)
 				continue
 			}
-			existing := raw.(uint32)
+
+			// these are the ones already in the tree
+			permissions := raw.(*Permissions)
+			existing := permissions.CapabilitiesBitmap
 
 			switch {
 			case existing&DenyCapabilityInt > 0:
 				// If we are explicitly denied in the existing capability set,
 				// don't save anything else
 
-			case pc.CapabilitiesBitmap&DenyCapabilityInt > 0:
+			case pc.Permissions.CapabilitiesBitmap&DenyCapabilityInt > 0:
 				// If this new policy explicitly denies, only save the deny value
-				tree.Insert(pc.Prefix, DenyCapabilityInt)
+				pc.Permissions.CapabilitiesBitmap = DenyCapabilityInt
+				tree.Insert(pc.Prefix, pc.Permissions)
 
 			default:
 				// Insert the capabilities in this new policy into the existing
 				// value
-				tree.Insert(pc.Prefix, existing|pc.CapabilitiesBitmap)
+				pc.Permissions.CapabilitiesBitmap = existing | pc.Permissions.CapabilitiesBitmap
+				tree.Insert(pc.Prefix, pc.Permissions)
+			}
+
+			// look for a * in allowed parameters for the node already in the tree
+			if _, ok := permissions.AllowedParameters["*"]; ok {
+				pc.Permissions.AllowedParameters = make(map[string]struct{})
+				pc.Permissions.AllowedParameters["*"] = struct{}{}
+				goto CHECK_DENIED
+			}
+
+			// look for a * in allowed parameters for the path capability we are merging
+			if _, ok := pc.Permissions.AllowedParameters["*"]; ok {
+				pc.Permissions.AllowedParameters = make(map[string]struct{})
+				pc.Permissions.AllowedParameters["*"] = struct{}{}
+				goto CHECK_DENIED
 			}
+
+			if pc.Permissions.AllowedParameters == nil {
+				pc.Permissions.AllowedParameters = make(map[string]struct{})
+			}
+			// Merge allowed parameters
+			for key, value := range permissions.AllowedParameters {
+				// Add new parameter
+				pc.Permissions.AllowedParameters[key] = value
+			}
```

Comment:
The three if blocks above should be merged like this (using your current variable naming): 
```go
if len(permissions.AllowedParameters) > 0 {
    if pc.Permissions.AllowedParameters == nil {
        pc.Permissions.AllowedParameters = make(map[string]struct{}{}, len(permissions.AllowedParameters))
    }
    if _, ok := pc.Permissions.AllowedParameters["*"]; ok {
        goto CHECK_DENIED
    }
    for key, value := range permissions.AllowedParameters {
        if key == "*" {
            pc.Permissions.AllowedParameters = map[string]struct{}{
                "*": struct{}{},
            }
            goto CHECK_DENIED
        }
        pc.Permissions.AllowedParameters[key] = value
    }
}
```
If you disagree, let me know :-)